### PR TITLE
#271 issue: Add call API function to general agent

### PR DESCRIPTION
**Pull request by vishu1089**
PR for #271

### CODE REVIEW

File: prediction_market_agent/agents/microchain_agent/call_api.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import requests
+
+class CallAPI:
```
Comment (kongzii) [inline prediction_market_agent/agents/microchain_agent/call_api.py:3]: CallAPI needs to inherit from Function, [for example as here](https://github.com/gnosis/prediction-market-agent/blob/5304d2cb043072407627e5e5323436c45b74d052/prediction_market_agent/agents/microchain_agent/memory_functions.py#L13).

File: prediction_market_agent/agents/microchain_agent/call_api.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    def __init__(self) -> None:
+        """
+        Initializes the class instance.
+        """
+        super().__init__()
```
Comment (kongzii) [inline prediction_market_agent/agents/microchain_agent/call_api.py:8]: This whole `__init__` method can be removed, as nothing goes on there.

File: prediction_market_agent/agents/microchain_agent/call_api.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    @property
+    def description(self) -> str:
+        """
+        A property method that returns a description of the function, which can be used to make arbitrary API calls to any endpoint with any parameters. It supports different HTTP methods (GET, POST, etc.) and can include headers, query parameters, and request bodies.
+        """
+        return (
+            "We can use this function to make arbitrary API calls to any endpoint with any parameters. "
```
Comment (kongzii) [inline prediction_market_agent/agents/microchain_agent/call_api.py:16]:
```suggestion
            "Use this function to make arbitrary API calls to any endpoint with any parameters. "
```

File: prediction_market_agent/agents/microchain_agent/call_api.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+            "It supports different HTTP methods (GET, POST, etc.) and can include headers, "
+            "query parameters, and request bodies."
+        )
+
+    @property
+    def example_args(self) -> list[str]:
+        """
+        Returns a list of strings representing the example arguments for the API call.
+
+        :return: A list of strings representing the example arguments for the API call.
```
Comment (kongzii) [inline prediction_market_agent/agents/microchain_agent/call_api.py:26]:
```suggestion
```

No need to duplicate comments

File: prediction_market_agent/agents/microchain_agent/call_api.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        :rtype: list[str]
```
Comment (kongzii) [inline prediction_market_agent/agents/microchain_agent/call_api.py:27]: We use python type annotations, so no need for this
```suggestion
```

File: prediction_market_agent/agents/microchain_agent/call_api.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        """
+        return ["method", "url", "params", "data", "headers"]
+
+    def __call__(self, method: str, url: str, params: dict = None, data: dict = None, headers: dict = None) -> str:
```
Comment (kongzii) [inline prediction_market_agent/agents/microchain_agent/call_api.py:31]:
```suggestion
    def __call__(self, method: str, url: str, params: dict[str, Any] | None = None, data: dict[str, Any] | None = None, headers: dict[str, Any] | None = None) -> str:
```

File: prediction_market_agent/agents/microchain_agent/call_api.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+        Returns:
+            str: The response string if the request was successful, or the error message if an exception occurred.
+        """
+        try:
+            response = requests.request(method, url, params=params, data=data, headers=headers)
+            response.raise_for_status()
+            return response.text
+        except requests.exceptions.RequestException as e:
+            return {"success": False, "error": str(e)}
```
Comment (kongzii) [inline prediction_market_agent/agents/microchain_agent/call_api.py:50]: Could you remove the try-except block please, other tools will throw an exception as well, so to not fail silently.

**Comment by vishu1089** [comment]
Super helpful!


Will take it as soon as possible!

File: prediction_market_agent/agents/microchain_agent/call_api.py (hunk 6 above)
Comment (kongzii) [reply prediction_market_agent/agents/microchain_agent/call_api.py:31]: Oh, for this to work, you will also need to add `from typing import Any` at the top of the file.

File: prediction_market_agent/agents/microchain_agent/call_api.py (hunk 2 above)
Comment (vishu1089) [reply prediction_market_agent/agents/microchain_agent/call_api.py:8]: okay

**Comment by vishu1089** [comment]
review changes done, please check

**Comment by vishu1089** [comment]
updated the branch with main

**Comment by kongzii** [comment]
@vishu1089 Thanks! Mostly perfect, but I pushed a few small tweaks (I didn't know what's wrong, and at the time I figured it out, it was basically done 😄 )

Tweaks include:

1. Run the code with black, autoflake and isort formatters ([see our pipeline](https://github.com/gnosis/prediction-market-agent/blob/918cf8ddfb870568b65d9bf4b5a719ed2521315b/.github/workflows/python_ci.yaml#L62))
2. Add the new Function tool to the agent, so he can use it
3. Fix the tool, because it didn't work with Microchain framework if input parameters were dictionaries, it need to be a string and then converted to the dictionary inside of the tool. (please, t[est out new functions using the app I suggested in the issue](https://github.com/gnosis/prediction-market-agent/issues/271#issuecomment-2188579007))
4. For some reason, microchain framework doesn't like typing using `|`, so I switched it to the old school `Optional[...]` instead.

I'll wait for the CI to pass and merge if all is okay.

Thanks for the help!

Screenshot by Dropbox Capture

**Review by kongzii — APPROVED** [review]

**Review by kongzii — APPROVED** [review]

**Comment by kongzii** [comment]
Some tests failed, but it's unrelated to this PR. Seems like secrets aren't available in PRs opened from forks (which makes sense). We need to look into how to fix that, so merging this now (cc @evangriffiths @gabrielfior ):

Screenshot by Dropbox Capture

**Pull request merged**